Check if PointsMaterial map is a WebGLRenderTarget.

vanruesc opened the pull request:
When using a WebGLRenderTarget as a map for the PointsMaterial, the following warning will be logged to the console every frame: _THREE.WebGLRenderTarget: .repeat is now .texture.repeat._

981e0a2 fixes this by checking if the given map is a WebGLRenderTarget.

2d4b856 takes care of a small TODO note. The canvas' clientHeight property doesn't need to be divided by 2 every frame. Now its stored as a local variable and will only be updated when the setSize method is called.

vanruesc committed:
Check if points material map is a render target.

Using a WebGLRenderTarget as a color map for the PointsMaterial results
in console spam (THREE.WebGLRenderTarget: .repeat is now
.texture.repeat.).

vanruesc committed:
Cache the half client height for the points scale.

Only update the half client height when setSize is called.

WestLangley commented:
If I recall correctly, `WebGLRenderTarget` used to be texture. Now it _has_ a `texture`.

It seems that the code pattern we should require now is
```javascript
material.map = renderTarget.texture; // not material.map = renderTarget;
```

which, as far as I know, works.

This will simplify the code a bit -- and, IMHO, makes more sense.

vanruesc commented:
@WestLangley You're right; when I first ran into this issue, I simply used the texture property of the render target directly which worked just fine. However, all the post processing examples for instance directly use their render targets as textures. That's why I decided to follow that style.

Anyway, I support the proposed code pattern.

Btw: the WebGLRenderer already contains code that treats render target as special cases: https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L2002-L2006.
 I used that as reference for this PR.

WestLangley commented:
>Btw: the WebGLRenderer already contains code that treats render target as special cases:

Right. And that special case can be removed.

I expect the `EffectComposer` changes will cause a lot of breakage, so it is up to @mrdoob if he wants to go this route.

mrdoob commented:
Yep, I also support @WestLangley suggested code pattern. I think the code clean up is worth the breakage.

mrdoob commented:
Anyone up for doing the refactoring?

vanruesc commented:
I'd be glad to do that.

Just to be clear:

- you want the WebGLRenderer to not treat WebGLRenderTargets as special cases anymore when they are used as maps (like [here](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L2515-L2527))
- and all examples that currently use render targets as textures should be refactored to use the **.texture** property instead

Right? I'm not sure where the breakage would happen with this refactoring, though :confused:

WestLangley commented:
>Right? 

Right.

>I'm not sure where the breakage would happen with this refactoring, though

It will break user's code. In which case it would be a good idea to test for `WebGLRenderTarget/Cube` and provide a warning to the user. The code should self-correct.

vanruesc committed:
Revert "Check if points material map is a render target."

This reverts commit 981e0a2d1837eac2a5de764c5b2609b1c7a255b6.

vanruesc committed:
Promote using the texture property of render targets.

Shadow maps are render targets; shadow as well as shadow.map can both be
null. EnvMaps need a special flag for the flipEnvMap check.

vanruesc committed:
Revert "Cache the half client height for the points scale."

This reverts commit 2d4b856ea60deffcdf20793a4a969e99a3f29f0b.

vanruesc committed:
Use .texture of current render target for encoding.

vanruesc committed:
Add renderTargetCube flag for flipEnvMap check.

vanruesc committed:
Code style.

vanruesc committed:
Use .texture of input render targets directly.

vanruesc committed:
Better stats.

vanruesc commented:
Alright, I'm done.

All examples now use the texture property of render targets that are used as inputs.

I noticed that the following examples were already broken before these changes:

- [Tone Mapping as a Post Processing Effect](http://threejs.org/examples/#webgl_shaders_tonemapping)
- [Dynamic Terrain](http://threejs.org/examples/#webgl_terrain_dynamic)

Additionally, the following examples seem to be broken in r76dev.

- materials / standard (physically based rendering)
- materials / envmaps / hdr

At least for me they fail to render the hdr envmaps (I checked with an untouched build of three-dev).

mrdoob commented:
@WestLangley looks good?

WestLangley commented:
Looks good.

@vanruesc Perhaps the `instanceof THREE.WebGLRenderTarget` tests can be made only when textures are assigned, rather than every render loop. If so, merge now; fix later.

vanruesc commented:
I think it'd be possible to modify all materials that define explicit [texture properties](https://github.com/mrdoob/three.js/blob/dev/src/materials/MeshLambertMaterial.js#L55) to use getters/setters instead. Checking for render targets could be done efficently this way and these changes could be reverted easily later on.

I'm not sure how to check custom shader texture uniforms efficiently, though, because they are [updated every frame](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L2497-L2529). Got any leads on this?

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2012,8 +2013,15 @@ THREE.WebGLRenderer = function ( parameters ) {
+
 		uniforms.envMap.value = material.envMap;
-		uniforms.flipEnvMap.value = ( material.envMap instanceof THREE.WebGLRenderTargetCube ) ? 1 : - 1;
+		uniforms.flipEnvMap.value = ( material.envMap && material.envMap.renderTargetCube ) ? 1 : - 1;
```
Shouldn't this be `&& material.envMap instanceof THREE.CubeTexture`? I wouldn't add `renderTargetCube` to `Texture` dynamically...

vanruesc replied:
The WebGLRenderTargetCube has a Texture instance because it extends WebGLRenderTarget. CubeTexture is not involved in this check any way. 

To replicate the original logic here, I had to resort to this awkward flag because there is no other way to know whether the texture at hand belongs to a WebGLRenderTargetCube.

mrdoob replied:
I see... I think we should figure out how to fix that first...

mrdoob commented:
Hmmm, this PR ended up having too many different changes. The refactoring is great but can't be merged because other things are not ready 😕

vanruesc commented:
Yea, I agree. It's kind of messed up at this point.

What do you think about the changes in the examples? 85a7aef1 should work with the current version of three right away. I could make a fresh PR out of that.

I'm a bit overwhelmed by the density of the WebGLRenderer class and I don't have enough time to dig through it to find a better solution. All I can offer right now is to open a new issue to outline the identified problem and keep hold of the details.

mrdoob commented:
> What do you think about the changes in the examples? 85a7aef1 should work with the current version of three right away. I could make a fresh PR out of that.

That would be great!

vanruesc commented:
Continued in #8615.

tschw commented:
Verdammt! Noch so'n irrer Deutscher der's nicht lassen kann den Renderer aufzuräumen... :)
[ EN: Damn! Yet 'nother Germaniac who can't resist to tidy up the renderer... :) ]

:+1: I appreciate your efforts.

vanruesc commented:
@tschw Well, it's definitely worth every effort! Three is where it's at :smile:
